api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java

```diff
@@ -104,20 +109,58 @@ private String insertParameters(OperationData input, String n3Text) throws Input
 
 		EditN3GeneratorVTwo gen = new EditN3GeneratorVTwo();
 
+		List<String> variables = getVariables(n3Text);
+
 		//Had to convert String to List<String> to be compatible with EditN3GeneratorVTwo methods
 		List<String> n3WithParameters = Arrays.asList(n3Text);
 
-		//region Substitute IRI variables
-		Map<String, List<String>> parametersToUris = requiredParams.getUrisMap(input);
-
-		gen.subInMultiUris(parametersToUris, n3WithParameters);
-		//endregion
-
-		//region Substitute other (literal) variables
-		Map<String, List<Literal>> parametersToLiterals = requiredParams.getLiteralsMap(input);
+		Map<String, List<String>> uriParametersToValues;
+		try {
+			uriParametersToValues =
+					variables.stream()
+							.map(variable -> variable.substring(1))
+							.distinct()
+							.filter(paramName -> requiredParams.get(paramName).getRDFDataType().getURI().equals(ANY_URI))
+							.collect(Collectors.toMap(
+									paramName -> paramName,
+									paramName -> Arrays.asList(input.get(paramName))
+							));
+		} catch (Exception e){
+			throw new InputMismatchException("Some N3 template variables dont have corresponding input parameters " +
+					"with values that should be inserted in their place.");
+		}
+		gen.subInMultiUris(uriParametersToValues, n3WithParameters);
 
-		gen.subInMultiLiterals(parametersToLiterals, n3WithParameters);
-		//endregion
+		Map<String, List<Literal>> literalParametersToValues;
+		try {
+			literalParametersToValues =
+					variables.stream()
+							.map(variable -> variable.substring(1))
+							.distinct()
+							.filter(paramName -> !requiredParams.get(paramName).getRDFDataType().getURI().equals(ANY_URI))
+							.collect(Collectors.toMap(
+									paramName -> paramName,
+									paramName -> Arrays.asList(ResourceFactory.createTypedLiteral(
+											input.get(paramName)[0],
+											requiredParams.get(paramName).getRDFDataType()
+									))
+							));
+		} catch (Exception e){
+			throw new InputMismatchException("Some N3 template variables dont have corresponding input parameters " +
+					"with values that should be inserted in their place.");
+		}
+		gen.subInMultiLiterals(literalParametersToValues, n3WithParameters);
+//		//region Substitute IRI variables
+//		Map<String, List<String>> parametersToUris = requiredParams.getUrisMap(input);
+//
+//		gen.subInMultiUris(parametersToUris, n3WithParameters);
+//		//endregion
+//
+//		//region Substitute other (literal) variables
+//		Map<String, List<Literal>> parametersToLiterals = requiredParams.getLiteralsMap(input);
+//
+//		gen.subInMultiLiterals(parametersToLiterals, n3WithParameters);
+//		//endregion
```

Comment:
I suggest removing those lines, if @VeljkoMaksimovic don't think it might be useful again in the future.

Reply:
I would leave them here for now, just until we decide whether the right place for substituting variables with concrete values is inside N3Template class (where they are currently) or inside Parameters class (where these commented out lines are). @litvinovg